### PR TITLE
AWS: make jittered_backoff API parameter configurable

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/cloudformation.py

### CODE REVIEW

**Comment by tedder** [comment]
I'd prefer the parameters to be prefixed to keep them logically (and lexicographically) together, since they are sort of "meta" parameters, not part of AWS's API itself. Perhaps "backoff_retries", "backoff_delay", etc.

That's just my opinion though, wondering what you think about that.

**Comment by Akasurde** [comment]
I am also OK with appending `backoff_` as in front of every parameter. I am also open to suggestions.

**Comment by s-hertel** [comment]
I like the `backoff_` prefix. I also think the options should be consistent when implemented in new modules, so documenting the choice in https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#api-throttling-and-pagination would be good.

@willthames Do you have thoughts on this?

**Comment by Akasurde** [comment]
@tedder @s-hertel Could you please take another look into this ? Thanks.

File: lib/ansible/modules/cloud/amazon/GUIDELINES.md [hunk 1]
```diff
@@ -416,6 +416,13 @@ describe_instances(module.client('ec2'), InstanceIds=['i-123456789'])
 The call will be retried the specified number of times, so the calling functions
 don't need to be wrapped in the backoff decorator.
 
+You can also use customization for `retries`, `delay` and `max_delay` parameters used by
+`AWSRetry.jittered_backoff` API using module params. You can take a look into
+[cloudformation](/lib/ansible/modules/cloud/amazon/cloudformation.py) module for example.
+
+To make all Amazon modules uniform, please use module param as `backoff_retries` for `retries`,
+`backoff_delay` of `delay` and `backoff_max_delay` for `max_delay`.
```
Comment (tedder) [inline lib/ansible/modules/cloud/amazon/GUIDELINES.md]: of->for, or just change the whole thing to explain that the backoff_ prefix should be added, and list the three examples:

> To make all Amazon modules uniform, prefix the module param with `backoff_`, so `retries` becomes `backoff_retries` and likewise with `backoff_delay` and `backoff_max_delay`.

File: lib/ansible/modules/cloud/amazon/cloudformation.py [hunk 2]
```diff
@@ -119,6 +119,28 @@
     - Maximum number of CloudFormation events to fetch from a stack when creating or updating it.
     default: 200
     version_added: "2.7"
+  backoff_delay:
```
Comment (tedder) [inline lib/ansible/modules/cloud/amazon/cloudformation.py:122]: idle thought: can these be added with a doc fragment instead?

File: lib/ansible/modules/cloud/amazon/cloudformation.py [hunk 3]
```diff
@@ -119,6 +119,28 @@
...
+    required: False
+  backoff_retries:
+    description:
+    - Number of times to retry operation.
+    - AWS API throttling mechanism fails Cloudformation module so we have to retry a couple of times.
+    default: 10
+    version_added: "2.8"
+    type: int
+    required: False
+  backoff_max_delay:
```
Comment (tedder) [inline lib/ansible/modules/cloud/amazon/cloudformation.py]: pedantic, but here and in the module params below, keep the order consistent. Thankfully alphasort will group the 'delay' by each other: backoff_delay, backoff_max_delay, backoff_retries.

File: lib/ansible/modules/cloud/amazon/GUIDELINES.md (hunk 1 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/amazon/GUIDELINES.md]: Done.

File: lib/ansible/modules/cloud/amazon/cloudformation.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/amazon/cloudformation.py:122]: Not all module support this. Once they do we can move this to doc fragment.

Comment (tedder) [reply lib/ansible/modules/cloud/amazon/cloudformation.py:122]: I was thinking about a new doc fragment, like aws_jittered, but it's fine too. I'll approve it for the bot.

**Comment by tedder** [comment]
Hooray, `shipit`.

**Pull request merged**